### PR TITLE
Introduce the new stats interface

### PR DESCRIPTION
**Pull request by maiste**
* This introduces a new module `Metrics` to describe metrics
* It updates the old `stat.ml` interface to support this new style.

### CODE REVIEW

File: src/irmin/metrics.mli [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  mutable storage : 'b;
+}
+
+val value : ('a, 'b) data -> 'a
+val storage : ('a, 'b) data -> 'b
+
+type ('a, 'b) kind =
+  | Ref
+  | Value of ('a -> 'a)
+  | Storage of ('a -> 'b -> 'b)
```
Comment (samoht) [inline src/irmin/metrics.mli:35]: I'm not sure to understand what these types mean :-) Might be useful to add comments/examples here?

Comment (maiste) [reply src/irmin/metrics.mli:35]: Yes, sure :smile: For now, it's just a draft so that @Ngoguey42 can look at it, and we have a place to discuss it.
But the plan is to improve the interface and add some documentation as soon as we are sure about it.

Comment (maiste) [reply src/irmin/metrics.mli]: I updated the documentation.

File: src/irmin-pack/stats.ml [hunk 2]
```diff
@@ -56,91 +95,203 @@ module Find = struct
...
-let incr_inode_remove () = s.inode_remove <- s.inode_remove + 1
-let incr_inode_of_seq () = s.inode_of_seq <- s.inode_of_seq + 1
-let incr_inode_of_raw () = s.inode_of_raw <- s.inode_of_raw + 1
-let incr_inode_rec_add () = s.inode_rec_add <- s.inode_rec_add + 1
-let incr_inode_rec_remove () = s.inode_rec_remove <- s.inode_rec_remove + 1
-let incr_inode_to_binv () = s.inode_to_binv <- s.inode_to_binv + 1
-let incr_inode_decode_bin () = s.inode_decode_bin <- s.inode_decode_bin + 1
-let incr_inode_encode_bin () = s.inode_encode_bin <- s.inode_encode_bin + 1
+let report_find ~(location : Find.location) = Find.report ~location s.finds
+let refresh_index () = Index.report s.index
```
Comment (maiste) [inline src/irmin-pack/stats.ml:269]:
```suggestion
let report_index () = Index.report s.index
```

File: src/irmin-pack/stats.mli [hunk 3]
```diff
@@ -35,44 +35,90 @@ module Find : sig
...
+      - [inode_encode_bin] is the number of [Bin] to [Compress] conversions;
+      - [inode_decode_bin] is the number of [Compress] to [Bin] conversions; *)
+
+  type stat
+
+  val retrieve : stat -> t
+end
+
+module Index : sig
+  type t = Index.Stats.t = {
```
Comment (maiste) [inline src/irmin-pack/stats.mli:95]:
```suggestion
  type t = Index.Stats.t = private {
```

File: src/irmin-pack/stats.mli [hunk 4]
```diff
@@ -35,44 +35,90 @@ module Find : sig
...
-      that is done modifying inodes;
-    - [inode_to_binv] is the number of [Inode.Bin.v] built;
-    - [inode_encode_bin] is the number of [Bin] to [Compress] conversions;
-    - [inode_decode_bin] is the number of [Compress] to [Bin] conversions; *)
+type t = { finds : Find.stat; pack : Pack.stat; index : Index.stat }
 
 val reset_stats : unit -> unit
 val get : unit -> t
 val report_find : location:Find.location -> unit
+val refresh_index : unit -> unit
```
Comment (maiste) [inline src/irmin-pack/stats.mli:121]:
```suggestion
val report_index : unit -> unit
```

File: src/irmin/metrics.mli [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+(** The [Metrics] module defines primitives to handle metrics inside of Irmin.
+    Its purpose is to decouple the metrics definition from the data
+    manipulation. *)
+
+type origin = ..
+(** An extensible type to get the location of the definition. *)
+
+type origin +=
+  | Origin_default  (** Default {!origin} for the {!data} definition. *)
```
Comment (samoht) [inline src/irmin/metrics.mli:25]: Maybe just call it `Default`?

File: src/irmin/metrics.mli [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+type 'a metric = { name : string; uid : int; ty : 'a Repr.t }
+(** {!metric} caracterizes the type of the metric that is manipulated. *)
```
Comment (samoht) [inline src/irmin/metrics.mli:28]: Would be better to make it abstract - easier to extend in the future if needed.

File: src/irmin/metrics.mli [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+type ('a, 'b) data = private {
+  uid : int;
+  name : string;
+  origin : origin;
+  metric : 'a metric;
+  mutable v : 'a;
+  mutable storage : 'b;
+}
```
Comment (samoht) [inline src/irmin/metrics.mli:37]: Same here, could you keep it abstract? It'll be easier to extend in the future.

File: src/irmin/metrics.mli [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+val value : ('a, 'b) data -> 'a
+(** The value field of a {!data} record. *)
+
+val storage : ('a, 'b) data -> 'b
+(** The storage field of a {!data} record. *)
+
+(** {!kind} describes how the data will be handled by the {!produce} function.
+
+    - Ref: the value and the storage are not manipulate
```
Comment (samoht) [inline src/irmin/metrics.mli:48]: manipulate -> modified

File: src/irmin/metrics.mli (hunk 6 above)
Comment (maiste) [reply src/irmin/metrics.mli:28]: Yes, you are right!

File: src/irmin/metrics.mli (hunk 1 above)
Comment (samoht) [reply src/irmin/metrics.mli]: A good way to design a good API is actually to try to explain it with documentation :-) ie. always start by a `.mli`, give a few hints on the general idea behind the API and give a few explanations on how users are supposed to use it. Writing a few examples is always a good way to find the right function and types names and the proper API to compose them.

Comment (samoht) [reply src/irmin/metrics.mli]: (I am still not totally sure how to use the current API bu just looking at the mli - some the names do not makes much sense - for instance, why `Ref`?)

File: src/irmin/metrics.mli (hunk 7 above)
Comment (maiste) [reply src/irmin/metrics.mli:37]: I agree, but it requires to introduce two functions:
```ocaml
val set_value : ('a, 'b) data -> 'a -> unit
val set_storage : ('a, 'b) data -> 'b -> unit 
```

File: src/irmin-pack/dune [hunk 9]
```diff
@@ -1,7 +1,7 @@
 (library
  (public_name irmin-pack)
  (name irmin_pack)
- (libraries fmt irmin irmin.data logs lwt optint)
+ (libraries fmt irmin irmin.data index logs lwt optint)
```
Comment (Ngoguey42) [inline src/irmin-pack/dune:4]: `irmin-pack` shouldn't depend on `index` because `irmin-pack` shouldn't depend on unix.

File: src/irmin/metrics.mli (hunk 1 above)
Comment (maiste) [reply src/irmin/metrics.mli]: Thanks for your explanation! I add this to my "Good Practises Guide" :smile: 
Moreover, I'll add some examples / hints at the beginning of the code as you suggest.

The `kind` type is supposed to describe three types of behaviours:
- The first one does nothing by itself
- The second one applies the function to update the `value` but not the storage
- The last one alters the `storage` and not the value
Of course, it's still possible to add more with other behaviours.

File: src/irmin/metrics.mli [hunk 10]
```diff
@@ -14,65 +14,70 @@
...
+(** {!ty} is the {!Repr.ty} description of the type used by metrics. *)
 
-type ('a, 'b) data
-(** {!data} is the object that describes how a {!metric} is gathered and store. *)
+val repr : string -> 'a Repr.ty -> 'a repr
+(** [repr name t] generates a metric representation, named according to [name]. *)
 
-val value : ('a, 'b) data -> 'a
-(** The value field of a {!data} record. *)
+type ('a, 'b) metric
```
Comment (samoht) [inline src/irmin/metrics.mli:35]: Maybe rename this to `t`? You'd use `Irmin.Metrics.t` anyway.

File: src/irmin/metrics.mli [hunk 11]
```diff
@@ -14,65 +14,70 @@
...
-(** [define name ty] creates a {ty metric}. *)
+    It gives the possibility to handle the same metric in different ways. *)
+type ('a, 'b) modifier =
+  | Mutable of ('a -> unit)
+  | Value of ('a -> 'a)
+  | Mem of ('a -> 'b -> 'b)
 
-val init :
-  'a metric ->
+val create :
```
Comment (samoht) [inline src/irmin/metrics.mli:66]: `v` ? :p (it's short, but it's kind of consistent with the rest of the create function in the Irmin codebase)

File: src/irmin/metrics.mli [hunk 12]
```diff
@@ -14,65 +14,70 @@
...
-(** [produce kind ~f data] updates the data by taking in consideration [kind]
-    and [f] if given as an argument. The purpose of [f] is to preupdate the data
-    if it requires an outside methode. It is mostly used with the [Ref] {!kind}. *)
-
-val consume : ('a, 'b) data -> converter:(('a, 'b) data -> 'c) -> 'c
-(** [consume ~converter data] returns the result from a {!data}. [converter] is
-    requires to describe how the result must be produced. [Fun.id] can be used
-    to only export the {!data}. *)
+  ('a, 'b) metric
+(** [create repr ~origin ~name ~defaut ~storage ()] create a new {!metric}. *)
```
Comment (samoht) [inline src/irmin/metrics.mli:74]: Would be nice to explain what `storage` is supposed to be here. It's not super clear to me how to use this in practice.

File: src/irmin/metrics.mli [hunk 13]
```diff
@@ -14,65 +14,70 @@
...
+    - Mem f: apply f to the storage and the value and updates the storage
+      content.
 
-val define : string -> 'a Repr.t -> 'a metric
-(** [define name ty] creates a {ty metric}. *)
+    It gives the possibility to handle the same metric in different ways. *)
+type ('a, 'b) modifier =
+  | Mutable of ('a -> unit)
+  | Value of ('a -> 'a)
+  | Mem of ('a -> 'b -> 'b)
```
Comment (samoht) [inline src/irmin/metrics.mli:64]: I don't understand why this is called `Mem` here?

File: src/irmin/metrics.ml [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+
+let update : type a b. (a, b) modifier -> (a, b) metric -> unit =
+ fun kind m ->
+  match kind with
+  | Mutable f -> f m.v
+  | Value f -> m.v <- f m.v
+  | Mem f -> m.storage <- f m.v m.storage
+
+let export : type a b c. (a, b) metric -> converter:((a, b) metric -> c) -> c =
+ fun d ~converter -> converter d
```
[the remark this answers is not in the document]
Comment (samoht) [reply src/irmin/metrics.ml:74]: (or do you expect to extend this in the future with something a bit more complex?)

File: src/irmin/metrics.mli (hunk 12 above)
Comment (maiste) [reply src/irmin/metrics.mli:74]: Sure I'll improve the documentation.

To summarize a bit here, it is supposed to allow us to do some histograms at some point by keeping the previous values in memory. As it needs to be easy to define new ones, what storage is, is not specified. Moreover, in this PR, it won't be used as we only want to keep track of the last value (`v`)

File: src/irmin/metrics.mli (hunk 13 above)
Comment (samoht) [reply src/irmin/metrics.mli:64]: How do prometheus / influxdb are calling this? Would be great to stick with the well-known names if possible (to not confuse users even more than with the current Irmin API :p)

Comment (samoht) [reply src/irmin/metrics.mli:64]: Maybe just reuse https://github.com/mirage/prometheus/blob/master/src/prometheus.mli#L13-LL17 ?

Comment (samoht) [reply src/irmin/metrics.mli:64]: Or just "counter"

File: src/irmin/metrics.ml (hunk 14 above)
[the remark this answers is not in the document]
Comment (samoht) [reply src/irmin/metrics.ml:74]: It's hard to say without seeing what the rest of the API looks like. I suspect most users will pick a specific converter (ie. only Prometheus) so I don't think that iterating over one metric is a very common operation. Having a full example would probably help us finding the right API :-)

File: src/irmin/metrics.mli (hunk 12 above)
Comment (maiste) [reply src/irmin/metrics.mli:74]: It's not an Irmin store. And this is not intended to be used by users, only by Irmin devs. The goal is to have a generic way to define metrics and export them in different ways. That's why its called storage and not store, to avoid the confusion :)

File: src/irmin/metrics.mli (hunk 13 above)
Comment (maiste) [reply src/irmin/metrics.mli:64]: As the goal is to be general, I don't know if it worths sticking to closely to Prometheus description...

File: src/irmin/metrics.ml (hunk 14 above)
[the remark this answers is not in the document]
Comment (maiste) [reply src/irmin/metrics.ml:74]: It's not intensed to be used by users. And the best example we have is the work on `stats.ml`.

File: src/irmin/metrics.mli (hunk 13 above)
Comment (maiste) [reply src/irmin/metrics.mli:64]: A suggestion by @Ngoguey42 was to call it `Acc` as it accumulates data.

Comment (Ngoguey42) [reply src/irmin/metrics.mli:64]: I would reword `metrics`, `storage`, `kind` and the 3 constructors that way:

`Metrics.t` contains a `state`. There are 3 `update_mode`s for the state: `Mutate`, `Replace` and `Accumulate`.

File: src/irmin/metrics.mli [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+    A {!metric} can be modified in different ways, depending on the {!modifier}. *)
+
+type origin = ..
+(** An extensible type to get the location of the definition. *)
+
+type 'a repr
+(** {!ty} is the {!Repr.ty} description of the type used by metrics. *)
+
+val repr : string -> 'a Repr.ty -> 'a repr
+(** [repr name t] generates a metric representation, named according to [name]. *)
```
Comment (Ngoguey42) [inline src/irmin/metrics.mli:33]: From the `Metrics` user point of view, is the only purpose of a `_ repr` to be passed to `create`? If true then why not inline the `string` and `a Repr.t` parameters in the `create` function and get rid of the `_ repr` type?

File: src/irmin/metrics.mli [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+
+type ('a, 'b) metric
+(** {!metric} is the object that describes how a {!metric} is gathered and
+    store. *)
```
Comment (Ngoguey42) [inline src/irmin/metrics.mli:37]: Can you describe the 2 types parameter here?

File: src/irmin/metrics.ml [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+let repr : type a. string -> a Repr.t -> a repr =
+ fun name ty -> { name; uid = uid (); ty }
+
+type ('a, 'b) metric = {
+  uid : int;
+  name : string;
+  origin : origin option;
+  repr : 'a repr;
+  mutable v : 'a;
+  mutable storage : 'b;
```
Comment (Ngoguey42) [inline src/irmin/metrics.ml:41]: Would it make sense to merge `v` and `storage` into a single `state` field?

**Review by Ngoguey42 — COMMENTED** [review]
My comments may contradict one another, sorry about that!

File: src/irmin/metrics.ml (hunk 14 above)
[the remark this answers is not in the document]
Comment (maiste) [reply src/irmin/metrics.ml:74]: In the function will be:
```ocaml
val export : ('a, 'b) data -> 'a * 'b
```
It makes more sense according to the discussion.

**Comment by maiste** [comment]
After a discussion with @Ngoguey42, I have simplified the interface. There are three purposes with the `Metrics` module and this PR:
1. Having a unified way to describe metrics in the `Irmin` code to make it easy to use.
2. Give the ability to create a `Metrics.t` so we can stop relying on a global reference and use a "formal type". As multicore is coming soon in Irmin, we need to be able to create `Store` local stats.
3. Thanks to the unified interface, it will be easy to write a `transform: Metrics.t -> Prometheus.t` function in the Tezos codebase.

I have pushed force with the new version, so it's ready for another review turn :+1:

File: src/irmin-pack/stats.mli [hunk 18]
```diff
@@ -14,67 +14,56 @@
...
 
-  val cache_misses : t -> int
-end
+      - [finds] stores the total number of calls to [S.find], and tracks the
+        source locations of successful finds (i.e. whether the value was
+        recovered from the staging table, LRU, or the pack file);
+      - [appended_hashes] is the number of times a hash was appended, during
+        calls to [add];
+      - [appended_offsets] is the number of times an offset was appended, during
+        calls to [add];
```
Comment (icristescu) [inline src/irmin-pack/stats.mli:49]: you forgot to remove this part of the comment

File: src/irmin-pack/unix/stats.mli [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+
+  val export : stat -> t
+end
+
+type t = { pack_store : Pack_store.stat; index : Index.stat }
+
+val reset_stats : unit -> unit
+val get : unit -> t
+val report_pack_store : field:Pack_store.field -> unit
+val report_index : unit -> unit
```
Comment (icristescu) [inline src/irmin-pack/unix/stats.mli:75]: maybe it would be useful to add a comment for the two `report_`? 
to explain that `report_pack_store` [also increments the field](https://github.com/mirage/irmin/pull/1817/files#diff-5b830a017f27442c88e2cb46ac9147c90363357193589954ef831b3a2c9270c5R186) whereas the following `report` function does not; 
also that they are needed before a call to `get`, is that right?

File: src/irmin-pack/unix/stats.ml [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+        appended_hashes = _;
+        appended_offsets = _;
+        (* Total finds (hits + misses): *)
+        total;
+        (* In-memory hits: *)
+        from_staging;
+        from_lru;
+        from_pack_direct = _;
+        from_pack_indexed = _;
+      } =
```
Comment (icristescu) [inline src/irmin-pack/unix/stats.ml:108]:
```suggestion
  let cache_misses
      {
        (* Total finds (hits + misses): *)
        total;
        (* In-memory hits: *)
        from_staging;
        from_lru;
        _;
      } =
```
(but as you prefer)

**Review by icristescu — APPROVED** [review]
Thanks, this looks great :)

File: src/irmin-pack/unix/stats.mli (hunk 19 above)
Comment (maiste) [reply src/irmin-pack/unix/stats.mli:75]: You can still get it but it won't be filled with other values than the `initial_state` value.

File: src/irmin-pack/unix/stats.ml (hunk 20 above)
Comment (maiste) [reply src/irmin-pack/unix/stats.ml:108]: It's lighter thanks!

**Review by Ngoguey42 — APPROVED** [review]
Thank you for all these changes :)

File: src/irmin/metrics.mli [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+(** {!t} is the object that describes how a {!t} is gathered and store. The ['a]
+    parameter represents the type of the internal data. *)
+
+val state : 'a t -> 'a
+(** The internal state extracted from a {!t}. *)
+
+val set_state : 'a t -> 'a -> unit
+(** [set_state m v] updates the value in the {!t} object. *)
+
+(** {!update} describes how the data will be handled by the {!produce} function.
```
Comment (tomjridge) [inline src/irmin/metrics.mli:35]: `{!update_mode}` ?
